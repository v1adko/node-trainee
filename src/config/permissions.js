export default {
  PUBLIC: { role: 'public', value: 0 },
  USER: { role: 'user', value: 1 },
  ADMIN: { role: 'admin', value: 2 }
};
