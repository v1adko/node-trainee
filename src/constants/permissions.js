export default {
  PUBLIC: { value: 'public', priority: 0 },
  USER: { value: 'user', priority: 1 },
  MODERATOR: { value: 'moderator', priority: 2 },
  ADMIN: { value: 'admin', priority: 3 }
};
