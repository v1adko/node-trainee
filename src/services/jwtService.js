import jwt from 'jsonwebtoken';
import configJwt from '../config/jwt';

const { tokenSecret: secret, tokenExpiresIn: expiresIn } = configJwt;

const options = { expiresIn };

class JwtService {
  generateJwt = (user) => {
    const payload = {
      _id: user._id,
      role: user.role
    };
    return jwt.sign(payload, secret, options);
  };

  decoder = (token) => {
    const promise = new Promise((resolve, reject) => {
      jwt.verify(token, secret, (error, decoded) => {
        if (error) {
          reject(new Error('Failed to authenticate token.'));
        } else {
          resolve(decoded);
        }
      });
    });
    return promise;
  };
}

export default new JwtService();
