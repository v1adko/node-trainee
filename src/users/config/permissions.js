export default {
  CREATE_USER: 'createUser',
  READ_USER: 'readUser',
  UPDATE_USER: 'updateUser',
  DELETE_USER: 'deleteUser',
  GET_MY_DATE: 'getMyDate',
  UPDATE_MY_DATE: 'updateMyDate'
};
