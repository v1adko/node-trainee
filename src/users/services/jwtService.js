const jwt = require('jsonwebtoken');

const { secretTokkenWord: secret } = require('../config/jwt');

const permissionsConst = require('../config/permissions');

class JwtService {
  constructor() {
    this.generateJwt = (user) => {
      const payload = {
        _id: user._id,
        permissions: Object.values(permissionsConst)
      };
      const options = {
        expiresIn: '2h'
      };

      return jwt.sign(payload, secret, options);
    };
  }
}

module.exports = new JwtService();
