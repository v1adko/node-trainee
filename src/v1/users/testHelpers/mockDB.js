import userDao from '../dao';
import userMocks from './userMocks';
import permissions from '../../../constants/permissions';

const { getUsersFields, user: userFields, admin: adminFields } = userMocks;

class MockDB {
  constructor(DAO) {
    this.DAO = DAO;
  }

  createDefaultUsers = async (userCount = 3) => {
    const users = [];
    getUsersFields(userCount).forEach(async (element) => {
      const { username, password, role } = element;
      const user = await this.createUser(username, password, role);
      users.push(user);
    });
  };

  createDefaultUser = async () => {
    const { username, password, role } = userFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createDefaultAdmin = async () => {
    const { username, password, role } = adminFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createUser = async (username, password, role = permissions.USER.value) => {
    const createdUser = this.DAO.create(username, password, role);
    return createdUser;
  };

  getAll = async () => {
    const users = await this.DAO.getAll();
    return users;
  };

  cleanDB = async () => {
    await this.DAO.Model.remove();
  };
}

export default new MockDB(userDao);
