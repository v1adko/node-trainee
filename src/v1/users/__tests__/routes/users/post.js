import simulate from '../../../../../utils/tests-utils/request-helper';
import User from '../../../user-model';
import jwtService from '../../../../services/jwt-service';
import permissions from '../../../../../constants/permissions';
import mockDB from '../../../../../utils/tests-utils/mock-db';
import UserHelper from '../../../../../utils/tests-utils/test-user-fields';

const ROUTE = '/v1/users';

const {
  username,
  password,
  shortUsername,
  longtPassword30,
  invalidRole
} = UserHelper;
const body = { username, password };
let adminToken;

async function clean() {
  await mockDB.cleanDB();
  adminToken = null;
}

async function createAdminToken() {
  const admin = new User();
  admin.role = permissions.ADMIN.value;
  adminToken = jwtService.generateJwt(admin);
}

beforeAll(clean);
afterAll(mockDB.closeConnection);

describe.skip(`Test the ${ROUTE} path`, () => {
  beforeEach(createAdminToken);
  afterEach(clean);

  it('should create user and return it in response on POST method', async () => {
    const { id, username: name } = await simulate.post(
      ROUTE,
      200,
      body,
      adminToken
    );

    expect(id).toEqual(expect.any(String));
    expect(name).toBe(username);
  });

  it.skip('should not create user because user already exist', async () => {
    // TODO: Fix it. It should work, but it not. And it create two users with same username. Why and how it do this?
    await mockDB.createUser(username, password);
    const { message } = await simulate.post(ROUTE, 405, body, adminToken);

    expect(message).toMatchSnapshot();
  });

  it('should not create user because username less than 6 symbols', async () => {
    const invalidBody = { username: shortUsername, password };
    const { message } = await simulate.post(
      ROUTE,
      400,
      invalidBody,
      adminToken
    );

    expect(message).toMatchSnapshot();
  });

  it('should not create user because password more than 30 symbols', async () => {
    const invalidBody = { username, password: longtPassword30 };
    const { message } = await simulate.post(
      ROUTE,
      400,
      invalidBody,
      adminToken
    );

    expect(message).toMatchSnapshot();
  });

  it('should not create user because role is invalid', async () => {
    const invalidBody = { username, password, role: invalidRole };
    const { message } = await simulate.post(
      ROUTE,
      400,
      invalidBody,
      adminToken
    );

    expect(message).toMatchSnapshot();
  });
});
