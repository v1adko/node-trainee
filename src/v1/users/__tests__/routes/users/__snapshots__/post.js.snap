// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the /v1/users path should not create user because password more than 30 symbols 1`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): \\"password\\" length must be less than or equal to 30 characters long",
  "name": "RequestValidationError",
}
`;

exports[`Test the /v1/users path should not create user because password more than 30 symbols 2`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): \\"password\\" length must be less than or equal to 30 characters long",
  "name": "RequestValidationError",
}
`;

exports[`Test the /v1/users path should not create user because role is invalid 1`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): \\"role\\" must be one of [user, moderator, admin]",
  "name": "RequestValidationError",
}
`;

exports[`Test the /v1/users path should not create user because role is invalid 2`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): \\"role\\" must be one of [user, moderator, admin]",
  "name": "RequestValidationError",
}
`;

exports[`Test the /v1/users path should not create user because username less than 6 symbols 1`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): \\"username\\" length must be at least 6 characters long",
  "name": "RequestValidationError",
}
`;

exports[`Test the /v1/users path should not create user because username less than 6 symbols 2`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): \\"username\\" length must be at least 6 characters long",
  "name": "RequestValidationError",
}
`;
